juspay/hyperswitch
# feat(roles): Add blacklist for roles

## Type of Change



## Description

This PR adds ability to blacklist roles to stop JWTs after a role is changed.

### Additional Changes





## Motivation and Context

Closes #3793 

## How did you test it?

Postman.
1. Create a role
	```
	curl --location 'http://localhost:8080/user/role' \
	--header 'Content-Type: application/json' \
	--header 'Authorization: Bearer JWT of Admin' \
	--data '{
	  "role_name": "custom_role",
	  "groups": ["users_view"],
	  "role_scope": "merchant"
	}'
	```
	Response should be 200 OK
2. Get the list of roles
	```
	curl --location 'http://localhost:8080/user/role/list' \
	--header 'Authorization: Bearer JWT of Admin'
	```
	In the response you will find newly created role
	```
	[
	    {
	        "role_id": "role_of8p9odNmf2puQVhQemT",
	        "permissions": [
	            "MerchantAccountRead",
	            "UsersRead"
	        ],
	        "role_name": "custom_role",
	        "role_scope": "merchant"
	    }
	]
	```
3. Update a user in the merchant in that merchant account to this newly created role
	```
	curl --location 'http://localhost:8080/user/user/update_role' \
	--header 'Content-Type: application/json' \
	--header 'Authorization: Bearer JWT of Admin' \
	--data-raw '{
	    "email": "email of other user in same merchant account",
	    "role_id": "role_of8p9odNmf2puQVhQemT"
	}'
	```
	Response will be 200 OK.
4. Login as the user who got updated
	```
	curl --location 'http://localhost:8080/user/signin' \
	--header 'Content-Type: application/json' \
	--data-raw '{
	    "email": "email of user who got udpated",
	    "password": "password"
	}'
	```
	```
	{
	    "token": "JWT of user who got updated",
	    "merchant_id": "merchant_id",
	    "name": "user name",
	    "email": "email of user who got updated",
	    "verification_days_left": null,
	    "user_role": "role_of8p9odNmf2puQVhQemT"
	}
	```
5. Update the role that is created earlier using admin JWT
	```
	curl --location --request PUT 'http://localhost:8080/user/role/role_of8p9odNmf2puQVhQemT' \
	--header 'Content-Type: application/json' \
	--header 'Authorization: Bearer JWT of Admin' \
	--data '{
	  "groups": ["users_view", "users_manage"]
	}'
	```
	Response will be 200 OK.
6. Try to use the token which we got earlier for the user who got updated and you should get InvalidJWTToken error.
	```
	{
	    "error": {
	        "type": "invalid_request",
	        "message": "Access forbidden, invalid JWT token was used",
	        "code": "IR_17"
	    }
	}
	```


## Checklist
